### DISCUSSION TITLE
Using REST API CRUD to handle create, remove, update and destroy?

### DISCUSSION THREAD

**Discussion by Jarrodsz** [remix-run/remix#4478] [Q&A]
Coming from a Ruby on Rails background Im used to CRUD actions based on POST,GET, etc.

I saw some implementations for this with Prisma.
there documentation lists several example repro's but not for remix currently.

For example this one:

https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-express

Would it be a good idea to have crud actions to handle my form submissions etc?
Or is this totally against remix conventions?

I know there is a file based router system that uses
/product/$id to display what used to be the show action.

**question**
Is it to be advised to use the route implementation in remix docs to process all form actions inside the action ?
So I could dry up the create, remove, update, destroy actions from a single route? and if so, how would the file based route located at /product/$id come into the equation.

```import * as React from "react";
import type { ActionFunction } from "@remix-run/node"; // or cloudflare/deno

export const action: ActionFunction = async ({request}) => {
  switch (request.method) {
    case "POST": {
      /* handle "POST" */
    }
    case "PUT": {
      /* handle "PUT" */
    }
    case "PATCH": {
      /* handle "PATCH" */
    }
    case "DELETE": {
      /* handle "DELETE" */
    }
  }
};


export default function SomePage() {

  return (
    <>
      crud
    </>
  );
}```

**Comment by machour** [comment]
PHPster here 👋🏼 

Yes, you could totally use actions to handle your CRUD operations (although technically speaking the R would be located in the `loader()` instead.

You can chose to handle all the operations in a single route, i.e. `/product/$id.tsx`. 

I would just not advise using `request.method` to distinguish operations, as this would only work with JS, since native forms can only handle `GET` or `POST`.
Instead, you could introduce an `intent` form element on your submit buttons:

```html
<button type="submit" name="intent" value="delete">DELETE</button>

<!-- ... -->

<button type="submit" name="intent" value="update">UPDATE</button>
```

and `switch` on `formData.get("intent")` instead of `request.method` in your action to execute the appropriate code.

---

From a routing point of view, you could also have this structure:

```
product
├---- create.tsx   <= C : /product/create
└── $id
    ├── index.tsx        <= R  : /product/$id
    ├── update.tsx       <= U  : /product/$id/update
    └── delete.tsx       <= D  : /product/$id/delete
```

and handle actions separately in each route

**Comment by Jarrodsz** [comment]
ah that is great (Y) thank you so much for confirming my initial thought and providing some context and address issues that where beyond my scope.

The examples in the prisma docs here:
```
https://www.prisma.io/docs/concepts/overview/prisma-in-your-stack/rest
```

List several implementations of a REST API build on several frameworks for example:
```
https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes/pages/api/post
```

I wonder if this is overkill or would be a really nice feature to have on remix also. 
Perhaps there is already an implementation/example on how to do rest api crud with remix. 
My searches in that regard did not yield any results.

**Reply by machour** [reply]
There's already everything needed in Remix to create a REST API, using [resource routes](https://remix.run/docs/en/v1/guides/resource-routes#creating-resource-routes). 👌🏼 

You could even handle HATEOAS by sending back the appropriate headers from your `loader()` or `action()`

Remix is more Ruby than Rails here, it's not opinionated, so you can do whatever you want with it.

**Reply by Jarrodsz** [reply]
How would u implement it with resource routes then?

Edit and new actions share a form page i assume one could do the same with s component.

From my rails background i do like to have an action for every vrud method.
